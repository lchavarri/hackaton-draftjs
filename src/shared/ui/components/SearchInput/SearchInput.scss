@import '../../assets/variables';

$search-clear-size: rem(30px);

.search {
  // SemanticUI override
  &.ui.input {
    > input[type='text'] {
      border-color: $border-color;
      border-radius: rem(15px);
      padding-right: $search-clear-size;

      &:focus,
      &:active {
        border-color: $hover;
        border-radius: rem(15px);
      }
    }

    > button.button {
      position: absolute;
      top: 0;
      right: rem(4px);
      width: $search-clear-size;
      height: $search-clear-size;
      padding: 0;
      margin: 0;
      background: transparent !important;

      &:hover {
        background: transparent !important;
      }

      > svg {
        width: rem(14px);
        height: rem(14px);
      }
    }

    .icon {
      & > svg {
        color: $menu;
      }
    }
  }
}
