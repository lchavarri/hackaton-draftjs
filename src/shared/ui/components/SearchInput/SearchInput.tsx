import React, { useState, useEffect, useRef } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSearch, faTimesCircle } from '@fortawesome/pro-light-svg-icons';
import { Input, Button } from 'semantic-ui-react';

import './SearchInput.scss';

type Props = {
  searchCallback: (term: string) => void;
  cancelCallback?: () => void;
  searchTerm: string;
  inputId: string;
  placeholder?: string;
  onBlur?: Function;
  isFocused?: boolean;
  onReset?: Function;
};

export default function SearchInput(props: Props) {
  const {
    searchCallback,
    cancelCallback = () => {},
    searchTerm,
    inputId,
    placeholder,
    onBlur = () => {},
    isFocused,
    onReset
  } = props;
  const input = useRef<any>(null);
  const [inputValue, setInputValue] = useState(searchTerm || '');

  useEffect(() => {
    if (inputValue !== searchTerm) {
      searchCallback(inputValue);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [inputValue]);

  useEffect(() => {
    if (inputValue !== searchTerm) {
      setInputValue(searchTerm);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [searchTerm]);

  useEffect(() => {
    if (isFocused && input && input.current) {
      input.current.focus();
    }
  }, [isFocused, input]);

  const onInputChange = (ev: React.ChangeEvent<HTMLInputElement>) => {
    setInputValue(ev.target.value);
  };

  const onKeyDown = (ev: KeyboardEvent) => {
    if (ev.key === 'Escape' && cancelCallback) {
      cancelCallback();
    }
  };

  const resetInput = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    e.stopPropagation();
    setInputValue('');
    if (input && input.current) {
      input.current.focus();
    }

    // Callback for parent component
    if (onReset) {
      onReset();
    }
  };

  const renderButton = () => {
    if (!inputValue || !inputValue.length) return null;

    return (
      <Button
        icon
        floated="right"
        onClick={resetInput}
        id={inputId + 'ResetInput'}
      >
        <FontAwesomeIcon
          icon={faTimesCircle}
          aria-label="Delete Criteria"
          className="icon"
        />
      </Button>
    );
  };

  return (
    <Input
      iconPosition="left"
      className="search clear"
      placeholder={placeholder || 'Search'}
      onChange={onInputChange}
      onBlur={onBlur}
      onKeyDown={onKeyDown}
      value={inputValue || ''}
      id={inputId + 'Term'}
      ref={input}
    >
      <i className="icon">
        <FontAwesomeIcon icon={faSearch} />
      </i>
      <input />
      {renderButton()}
    </Input>
  );
}
