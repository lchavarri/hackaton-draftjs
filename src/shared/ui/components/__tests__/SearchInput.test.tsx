import React from 'react';
import { shallow, mount, ReactWrapper } from 'enzyme';
import SearchInput from '../SearchInput/SearchInput';

describe('SearchInput component', () => {
  let component: ReactWrapper;
  let props: any;
  const searchCallback = jest.fn();

  const defaultProps = {
    inputId: 'test',
    searchCallback,
    searchTerm: 'Init'
  };
  const getComponent = () => <SearchInput {...props} />;
  const wrap = () => shallow(getComponent());
  const wrapDeep = () => mount(getComponent());

  beforeEach(() => {
    jest.restoreAllMocks();
    props = defaultProps;
    component = wrap();
  });

  it('should match component snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('should have the initial searchTerm setted', () => {
    props = { ...defaultProps, searchTerm: '123' };
    component = wrapDeep();

    const inputValue: string | number | string[] = component
      .find('input#testTerm')
      .prop('value');

    expect(inputValue).toEqual('123');
  });

  it('should not call searchCallback on initial load', () => {
    props = { ...defaultProps, searchTerm: '456' };
    component = wrapDeep();

    const inputValue: string | number | string[] = component
      .find('input#testTerm')
      .prop('value');

    expect(inputValue).toEqual('456');
    expect(searchCallback).not.toHaveBeenCalled();
  });

  it('should render the clear button is input has value', () => {
    component = wrapDeep();

    expect(component.exists('button#testResetInput')).toBe(true);
  });

  it('should not render the clear button is input is empty', () => {
    props = { ...defaultProps, searchTerm: '' };
    component = wrapDeep();

    expect(component.exists('button#testResetInput')).toBe(false);
  });

  describe('on input', () => {
    let input: ReactWrapper;
    beforeEach(() => {
      jest.restoreAllMocks();
      component = wrapDeep();
      input = component.find('input#testTerm');
      input.simulate('change', { target: { value: 'update' } });
    });

    it('should call searchCallback with the new value', () => {
      expect(searchCallback).toHaveBeenCalledWith('update');
    });
  });

  describe('on click clear', () => {
    beforeEach(() => {
      jest.restoreAllMocks();
      component = wrapDeep();
      component.find('button#testResetInput').simulate('click');
    });

    it('should delete previous input', () => {
      const inputValue: string | number | string[] = component
        .find('input#testTerm')
        .prop('value');

      expect(inputValue).toEqual('');
    });

    it('should call searchCallback with the empty string', () => {
      expect(searchCallback).toHaveBeenCalledWith('');
    });
  });
});
