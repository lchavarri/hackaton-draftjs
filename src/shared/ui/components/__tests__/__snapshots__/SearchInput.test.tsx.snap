// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput component should match component snapshot 1`] = `
<Input
  className="search clear"
  iconPosition="left"
  id="testTerm"
  onBlur={[Function]}
  onChange={[Function]}
  onKeyDown={[Function]}
  placeholder="Search"
  type="text"
  value="Init"
>
  <i
    className="icon"
  >
    <mockConstructor />
  </i>
  <input />
  <Button
    as="button"
    floated="right"
    icon={true}
    id="testResetInput"
    onClick={[Function]}
  >
    <mockConstructor
      aria-label="Delete Criteria"
      className="icon"
    />
  </Button>
</Input>
`;
